gflownet/utils/buffer.py

```diff
@@ -259,11 +259,47 @@ def make_data_set(self, config):
         )
         return df, {"x": samples, "energy": energies}
 
-    def compute_stats(self, data):
+    @staticmethod
+    def compute_stats(data):
         mean_data = data["energies"].mean()
         std_data = data["energies"].std()
         min_data = data["energies"].min()
         max_data = data["energies"].max()
         data_zscores = (data["energies"] - mean_data) / std_data
         max_norm_data = data_zscores.max()
         return mean_data, std_data, min_data, max_data, max_norm_data
+
+    @staticmethod
+    def select(data_dict: dict, n: int, mode: str = "permutation", rng=None):
+        if n == 0:
+            return []
+        samples = data_dict["x"]
+        # If the data_dict comes from the replay buffer, then samples is a dict and we
+        # need to keep its values only
+        if isinstance(samples, dict):
+            samples = list(samples.values())
+        if mode == "permutation":
+            assert rng is not None
+            samples = [samples[idx] for idx in rng.permutation(n)]
+        elif mode == "weighted":
+            if "rewards" in data_dict:
+                score = "rewards"
+            elif "energy" in data_dict:
+                score = "energy"
+            else:
+                raise ValueError(f"Data set does not contain reward or energy key.")
+            scores = data_dict[score]
+            # If the data_dict comes from the replay buffer, then scores is a dict and we
+            # need to keep its values only
+            if isinstance(scores, dict):
+                scores = np.fromiter(scores.values(), dtype=float)
+            indices = np.random.choice(
+                len(samples),
+                size=n,
+                replace=False,
+                p=scores / scores.sum(),
+            )
+            samples = [samples[idx] for idx in indices]
```

Comment:
I wondered which of those patterns was the fastest:

```
def f(l, n):
	indices = np.random.choice(len(l), size=n, replace=False)
	return [samples[i] for i in indices]

def g(l, n):
    return np.random.choice(l, size=n, replace=False).tolist()
```

`f` is ~2.5x faster than `g` for `len(l) = 2000` and `n = 1000` (and it gets even faster when `len(l)` increases.

Reply:
So I went with the faster choice! :v: